server/src/main/java/org/elasticsearch/index/mapper/VersionStringFieldMapper.java

```diff
@@ -0,0 +1,380 @@
+        this.ignoreMalformed = ignoreMalformed;
+        this.ignoreAbove = ignoreAbove;
+        this.nullValue = nullValue;
+        this.mode = mode;
+    }
+
+    @Override
+    public VersionStringFieldType fieldType() {
+        return (VersionStringFieldType) super.fieldType();
+    }
+
+    @Override
+    protected String contentType() {
+        return CONTENT_TYPE;
+    }
+
+    @Override
+    protected VersionStringFieldMapper clone() {
+        return (VersionStringFieldMapper) super.clone();
+    }
+
+    @Override
+    protected void parseCreateField(ParseContext context) throws IOException {
+        String versionString;
+        if (context.externalValueSet()) {
+            versionString = context.externalValue().toString();
+        } else {
+            XContentParser parser = context.parser();
+            if (parser.currentToken() == XContentParser.Token.VALUE_NULL) {
+                versionString = nullValue;
+            } else {
+                versionString =  parser.textOrNull();
+            }
+        }
+
+        if (versionString == null) {
+            return;
+        }
+
+        BytesRef encodedVersion = null;
+        try {
+            encodedVersion = encodeVersion(versionString, mode);
+        } catch (IllegalArgumentException e) {
+            if (ignoreMalformed.value()) {
+                context.addIgnoredField(name());
+                return;
+            } else {
+                throw e;
+            }
+        }
+        if (fieldType.indexOptions() != IndexOptions.NONE || fieldType.stored())  {
+            Field field = new Field(fieldType().name(), encodedVersion, fieldType);
+            context.doc().add(field);
+
+            if (fieldType().hasDocValues() == false && fieldType.omitNorms()) {
+                createFieldNamesField(context);
+            }
+        }
+
+        if (fieldType().hasDocValues()) {
+            context.doc().add(new SortedSetDocValuesField(fieldType().name(), encodedVersion));
+        }
+
+    }
+
+
+
+    @Override
+    protected void mergeOptions(FieldMapper other, List<String> conflicts) {
+        VersionStringFieldMapper mergeWith = (VersionStringFieldMapper) other;
+        if (mergeWith.ignoreMalformed.explicit()) {
+            this.ignoreMalformed = mergeWith.ignoreMalformed;
+        }
+        this.ignoreAbove = mergeWith.ignoreAbove;
+        this.nullValue = mergeWith.nullValue;
+        this.mode = mergeWith.mode;
+    }
+
+    @Override
+    protected void doXContentBody(XContentBuilder builder, boolean includeDefaults, Params params) throws IOException {
+        super.doXContentBody(builder, includeDefaults, params);
+
+        if (nullValue != null) {
+            builder.field("null_value", nullValue);
+        }
+
+        if (includeDefaults || ignoreMalformed.explicit()) {
+            builder.field("ignore_malformed", ignoreMalformed.value());
+        }
```

Comment:
this must support all options that are supported by the parser, so it looks like some of them are missing

Reply:
added output for "mode" for now